src/V3Const.cpp

```diff
@@ -3565,9 +3557,9 @@ class ConstVisitor final : public VNVisitor {
     TREEOPV("AstOneHot{$lhsp.width1}",          "replaceWLhs(nodep)");
     TREEOPV("AstOneHot0{$lhsp.width1}",         "replaceNum(nodep,1)");
     // Binary AND/OR is faster than logical and/or (usually)
-    TREEOPV("AstLogAnd{$lhsp.width1, $rhsp.width1, isTPure($lhsp), isTPure($rhsp)}", "AstAnd{$lhsp,$rhsp}");
-    TREEOPV("AstLogOr {$lhsp.width1, $rhsp.width1, nodep->isPure(), isTPure($lhsp), isTPure($rhsp)}", "AstOr{$lhsp,$rhsp}");
-    TREEOPV("AstLogNot{$lhsp.width1, isTPure($lhsp)}",  "AstNot{$lhsp}");
+    TREEOPV("AstLogAnd{$lhsp.width1, $rhsp.width1, $rhsp.isPure}", "AstAnd{$lhsp,$rhsp}");
+    TREEOPV("AstLogOr {$lhsp.width1, $rhsp.width1, $rhsp.isPure}", "AstOr{$lhsp,$rhsp}");
```

Comment:
Later optimizations may swap sides, so I think we should still require $lhsp.isPure.

Reply:
Swapping sides breaks short-circuiting. I think we should disable it in case of `||` and `&&`.

Reply:
Agreed, please add that if not already there.

Reply:
I wanted to do so, but I don't know which optimizations may swap operands. Do you know any such optimizations?

Reply:
```
    // Prefer constants on left, since that often needs a shift, it lets
    // constant red remove the shift
    TREEOP ("AstNodeBiCom{!$lhsp.castConst, $rhsp.castConst}",  "swapSides(nodep)");
    TREEOP ("AstNodeBiComAsv{operandAsvConst(nodep)}",  "replaceAsv(nodep)");
    TREEOP ("AstNodeBiComAsv{operandAsvSame(nodep)}",   "replaceAsv(nodep)");
    TREEOP ("AstNodeBiComAsv{operandAsvLUp(nodep)}",    "replaceAsvLUp(nodep)");
    TREEOP ("AstNodeBiComAsv{operandAsvRUp(nodep)}",    "replaceAsvRUp(nodep)");

```

Reply:
I suppose really AstOr/AstAnd shouldn't be AstNodeBiComAsv, but it's probably easier to leave them as such, add a comment to them in V3AstNodeExpr.h, and search where AstNodeBiCom is used and check isPure() there.

Reply:
Bitwise `or` and `and` are commutative and associative. I think it will be 100% correct to keep them as AstNodeBiComAsv.
When I was writing the first comment, I thought that optimizations may swap sides of AstLogOr and AstLogAnd. Now I know that you meant the optimizations performed on bitwise operations after the conversion. So maybe it would be better to disable the conversion if the LHS isn't pure.

Reply:
I did it. One test failed on clang due to a constant operand in `||`. It was the right operand, so it wasn't removed by optimizations. I think we can ignore such warnings, so I added a flag.